go-controller/pkg/ovn/policy.go

```diff
@@ -1027,59 +1030,88 @@ func (oc *Controller) createNetworkPolicy(np *networkPolicy, policy *knet.Networ
 	defer np.Unlock()
 	if np.deleted {
 		oc.processLocalPodSelectorDelPods(np, selectedPods...)
-		return
+		return nil
 	}
 
 	ops, err = libovsdbops.CreateOrUpdatePortGroupsOps(oc.nbClient, ops, pg)
 	if err != nil {
 		oc.processLocalPodSelectorDelPods(np, selectedPods...)
-		klog.Errorf(err.Error())
-		return
+		return fmt.Errorf("failed to create ops to add port to a port group: %v", err)
 	}
 
 	_, err = libovsdbops.TransactAndCheck(oc.nbClient, ops)
 	if err != nil {
 		oc.processLocalPodSelectorDelPods(np, selectedPods...)
-		klog.Errorf(err.Error())
-		return
+		return fmt.Errorf("failed to run ovsdb txn to add ports to port group: %v", err)
 	}
 	np.created = true
+	return nil
 }
 
 // addNetworkPolicy creates and applies OVN ACLs to pod logical switch
 // ports from Kubernetes NetworkPolicy objects using OVN Port Groups
-func (oc *Controller) addNetworkPolicy(policy *knet.NetworkPolicy) {
+func (oc *Controller) addNetworkPolicy(policy *knet.NetworkPolicy) error {
 	klog.Infof("Adding network policy %s in namespace %s", policy.Name,
 		policy.Namespace)
 
 	nsInfo, nsUnlock, err := oc.ensureNamespaceLocked(policy.Namespace, false, nil)
 	if err != nil {
-		klog.Errorf("Unable to ensure namespace for network policy: %s, namespace: %s, error: %v",
+		return fmt.Errorf("unable to ensure namespace for network policy: %s, namespace: %s, error: %v",
 			policy.Name, policy.Namespace, err)
-		return
 	}
 	_, alreadyExists := nsInfo.networkPolicies[policy.Name]
 	if alreadyExists {
 		nsUnlock()
-		return
+		// If this scenario happens something is wrong in our code, however if we return error here
+		// the NP will be retried infinitely to be added. Another option would be to fatal out here
+		// but that seems too aggressive
+		klog.Errorf("During add network policy, policy already found for %s/%s, this should not happen!",
+			policy.Namespace, policy.Name)
+		return nil
 	}
 	np := NewNetworkPolicy(policy)
 
 	if len(nsInfo.networkPolicies) == 0 {
 		err = oc.createDefaultDenyPGAndACLs(policy.Namespace, policy.Name, nsInfo)
 		if err != nil {
-			klog.Errorf(err.Error())
 			nsUnlock()
-			return
+			return fmt.Errorf("failed to create default port groups and acls for policy: %s/%s, error: %v",
+				policy.Namespace, policy.Name, err)
 		}
 	}
-	nsInfo.networkPolicies[policy.Name] = np
```

Comment:
yeah this was intentional so that on namespace update, it would be able to see the policy to update the ACL logging in the case where:
1. np is being created, reads what the acl logging is off the namespace
2. while its being created, someone updates the namepace logging
3. namespace hander checks what network policies exist (sees this one), and waits for NP thread to finish creating it, then updates it

This is all because we have to give up the NP lock while we create the NP, and the NP code isn't thread safe itself. To workaround the problem of potentially missing an acl logging update, I added the acl logging update code once we get the nsInfo lock back in the this function at lines 1107. It's not pretty, but it should work and be less risky than  https://github.com/ovn-org/ovn-kubernetes/pull/2797